File: LeetCode/Combinations.py
class Solution:
    # @return a list of lists of integers
    def combine(self, n, k):
        self.result = []
        self.n = n
        self.k = k
        buf = []
        self.findCombination(1, 0, buf)
        return self.result

    def findCombination(self, start, length, buf):
        if start > self.n+1:
            return
        if self.k > length+self.n-start +1:
            return
        # print (start,length,buf)
        if length == self.k:
            self.result.append(buf)
            return
        self.findCombination(start+1, length, buf)
        new_buf = buf[:]
        new_buf.append(start)
        self.findCombination(start+1, length+1, new_buf)
        return

if __name__ == '__main__':
    s = Solution()
    print s.combine(4,2)

File: LeetCode/Combinations.cpp
class Solution {
public:
    vector<vector<int> > combine(int n, int k) {
        vector<vector<int> > ans;
        vector<int> result;
        dfs(ans, result, n, 1, k);
        return ans;
    }
    void dfs(vector<vector<int> > &ans, vector<int> result, int n, int num, int k){
        if (k == 0){
            ans.push_back(result);
            return;
        }

        if (num <= n-k+1 && k > 0){
            for (int i = num; i <= n; i++){
                result.push_back(i);
                dfs(ans, result, n, i+1, k-1);
                result.pop_back();
            }
        }
    }
};